add compiler page (#33)
status-im/nim-style-guide: src/tooling.compiler.md

## Compiler `[tooling.compiler]`

### config.nims

> Use `config.nims` to harden your build by promoting selected warnings (and hints) to errors, forcing safer, cleaner code to pass compilation.

Reviewer: it might be worth pointing out here that config.nims from a library typically does not apply to "users" of the library _unless_ the library resides in a subfolder of the project using it - this is a common source of confusion.

Reviewer 2: That's an advantage, not a disadvantage, mostly: it means that the library's own CI system can catch more, but doesn't create unexpected downstream effects.

**Treat Selected Warnings as Errors**

Consider promoting certain warnings to errors to enforce stricter code quality:
```nim
switch("warningAsError", "UnreachableCode:on")
```

- `UnreachableCode:on` → Catches code that can never execute
- `UnreachableElse:on` → Detects else branches that are never reachable
- `UnusedImport:on` → Keeps the codebase clean by removing unused imports
- `UseBase:on` → Ensures correct use of base methods in inheritance
- `ResultShadowed:on` → Prevents accidentally redefining result, which can hide return values.

**Treat Selected Hints as Errors**

You can also promote certain hints to errors:
```nim
switch("hintAsError", "XCannotRaiseY:on")
```

> - `XCannotRaiseY:on` → Flags invalid exception declarations for cleaner interaces and simplier error handling

Reviewer 2: We've found `XCannotRaiseY` is one of the trickier ones to eliminate in any code which uses generics, templates, et cetera, empirically. The rest of these are comparatively straightforward. The issue is that some instantiations genuinely do raise those exceptions, and some don't. For codebases for which it works, yes, it's a good invariant to preserve.

Author: even with that in mind, it should still be as general recommendation?

same warning can also be issue with methods, one impelemation could rase X while other might not. but warning can be disabled for specific case .

Reviewer 2: It's still worth generally recommending it, yes. However, I'd distinguish here between the hints and warnings which
1. are fairly deterministically straightforward to satisfy in a local way in all the IFT Nim codebases I've yet encountered (`UnusedImport`, `UseBase`, `ResultShadowed`, `DuplicateModuleImport`, and `ConvFromXtoItselfNotNeeded`, for example, though not necessarily exhaustively);
2. those where it does come up semi-legitimately but only in kind of strange edge cases (`UnreachableCode`/`UnreachableElse`, typically because it's easier to template or macro-generate code which in certain circumstances results in this) but really should be resolvable and it generally does improve the codebase to do so; and
3. `XCannotRaiseY`, which I've found to be, in Nimbus, likely near-term [not worth attempting to resolve](https://github.com/status-im/nimbus-eth2/blob/07d88087d0886b71711287e1049cb3599b3b3322/config.nims#L169-L179):
```
#   1 nimbus-eth2/tests/consensus_spec/test_fixture_ssz_generic_types.nim(238, 28) Hint: 'sszCheck' cannot raise 'YamlConstructionError' [XCannotRaiseY]
#   1 nimbus-eth2/tests/consensus_spec/test_fixture_ssz_generic_types.nim(238, 51) Hint: 'sszCheck' cannot raise 'YamlParserError' [XCannotRaiseY]
#   1 nimbus-eth2/vendor/nim-testutils/testutils/moduletests.nim(17, 24) Hint: 'main' cannot raise 'CatchableError' [XCannotRaiseY]
#   2 nimbus-eth2/tests/consensus_spec/test_fixture_light_client_sync.nim(135, 20) Hint: 'loadTestMeta' cannot raise 'YamlConstructionError' [XCannotRaiseY]
#   2 nimbus-eth2/tests/consensus_spec/test_fixture_light_client_sync.nim(135, 43) Hint: 'loadTestMeta' cannot raise 'YamlParserError' [XCannotRaiseY]
#   2 nimbus-eth2/vendor/nim-toml-serialization/toml_serialization/reader.nim(213, 58) Hint: 'readValue' cannot raise 'IOError' [XCannotRaiseY]
#   3 nimbus-eth2/vendor/nim-toml-serialization/toml_serialization/reader.nim(369, 38) Hint: 'readValue' cannot raise 'SerializationError' [XCannotRaiseY]
#   3 nimbus-eth2/vendor/nim-toml-serialization/toml_serialization/reader.nim(369, 58) Hint: 'readValue' cannot raise 'IOError' [XCannotRaiseY]
#   4 nimbus-eth2/vendor/nim-serialization/serialization.nim(27, 86) Hint: 'readValue' cannot raise 'IOError' [XCannotRaiseY]
# 116 nimbus-eth2/vendor/nim-ssz-serialization/ssz_serialization.nim(51, 77) Hint: 'writeFixedSized' cannot raise 'IOError' [XCannotRaiseY]
switch("hint", "XCannotRaiseY:off")
```

Ideally, the goal here is to create `config.nims` snippet which can be reliably copy/pasted, and the codebase adapted to fit with minimal risk or external changes to the repository itself. This includes, for me, all of category (1), and arguably category (2). Category (3) verges on being counterproductive if it puts people off doing this at all or makes the entire exercise look broken/infeasible.

So yes, it's worth recommending, but in a way cognizant of the relative difficulties of achieving each.

Also, `nimbus-eth2` adds a few more to your existing list; perhaps worth including:
```
switch("warningAsError", "CaseTransition:on")
switch("warningAsError", "CStringConv:on")
switch("warningAsError", "ImplicitDefaultValue:on")
switch("warningAsError", "LongLiterals:on")
```
`CStringConv` is particularly important for basic safety/reliability and `CaseTransition` amounts to avoiding a long-deprecated feature of how case objects work in Nim, which is decreasingly well-supported over time and has had outright undefined behavior sometimes.

Reviewer 2: Of these, `CaseTransition` sometimes does come up via libraries, and can be case (2) or (3), so should be noted such. The other three are basically all locally, deterministically fixable, so category (1) of probably would be in the default copy/paste block.

## Compiler `[tooling.compiler]`

### config.nims

Use `config.nims` to harden your build by promoting selected warnings (and hints) to errors, forcing safer, cleaner code to pass compilation.

> Below is `config.nims` snippet which can be reliably copy/pasted; it converts compile time warnings and hints to errors.

Reviewer 2: ```suggestion
A `config.nims` snippet which can be copy/pasted follows; it converts certain compile-time warnings and hints to errors.
```

```nim
# turn warnings into compile errors
switch("warningAsError", "BareExcept:on")
switch("warningAsError", "CaseTransition:on")
switch("warningAsError", "CStringConv:on")
switch("warningAsError", "ImplicitDefaultValue:on")
switch("warningAsError", "LongLiterals:on")
switch("warningAsError", "ResultShadowed:on")
switch("warningAsError", "UnreachableCode:on")
switch("warningAsError", "UnreachableElse:on")
switch("warningAsError", "UnusedImport:on")
switch("warningAsError", "UseBase:on")

# turn hints into compile errors
switch("hintAsError", "ConvFromXtoItselfNotNeeded:on")
switch("hintAsError", "DuplicateModuleImport:on")
switch("hintAsError", "XCannotRaiseY:on")
```

> Note that if some of above does not work for your project, such as `XCannotRaiseY`, feel free to leave it out.

Reviewer 2: While there are some copyediting points (the only-tweaked-for-grammatical-correctness version is something like "Note that if some of above, such as `XCannotRaiseY`, do not work for your project, feel free to leave them out.") there's more of an issue here: it adopts a different tone than the rest of the guide with "feel free to ...". There's an informality/casualness which is a little tonally off from the style elsewhere.

(This is not to say that it should not note that one doesn't have to use the full set, but consistency in tone matters.)

Author: please suggest

Reviewer 2: > Some of these hints or warnings, such as `XCannotRaiseY`, might prove situationally impractical to enforce as errors; these can be omitted as required.

Captures a reasonable tone/balance/consistency with the rest of the style guide. Doesn't use a directional metaphor which is somewhat dicey with flexible web layouts. Keeps the somewhat abstracted-from-specific-project language that the guide generally uses, while still encouraging usage of as many as feasible.

> Note that if some of above does not work for your project, such as `XCannotRaiseY`, feel free to leave it out.

Author: ```suggestion

Some of these hints or warnings, such as` XCannotRaiseY`, might prove situationally impractical to enforce as errors; these can be omitted as required.
```
